### PR TITLE
Flipping the __sub__ method of ImageMultiHash

### PR DESCRIPTION
**Pull request by joshcoales**
and fixing divide by zero issue

### CODE REVIEW

File: imagehash.py [hunk 1]
```diff
@@ -390,10 +390,23 @@ def __eq__(self, other):
 			return False
 		return self.matches(other)
 
+	def __ne__(self, other):
+		if other is None:
+			return False
```
Comment (JohannesBuchner) [inline imagehash.py:395]: should this be False?

File: imagehash.py [hunk 2]
```diff
@@ -390,10 +390,23 @@ def __eq__(self, other):
...
+		return not self.matches(other)
+
 	def __sub__(self, other, hamming_cutoff=None, bit_error_rate=None):
 		matches, sum_distance = self.hash_diff(other, hamming_cutoff, bit_error_rate)
+		max_difference = len(self.segment_hashes)
+		if matches == 0:
+			return max_difference
 		max_distance = matches * len(self.segment_hashes[0])
-		return matches + (1 - (sum_distance / max_distance))
+		tie_breaker = 1 - (sum_distance / max_distance)
```
Comment (JohannesBuchner) [inline imagehash.py:404]: Please check all " / ", this project is still supporting Python 2, which doesn't convert int/int divisions to floats automatically.

**Review by JohannesBuchner — COMMENTED** [review]
Please also run the tests and check the test coverage (travis.yml should have the commands).

**Comment by joshcoales** [comment]
> Please also run the tests and check the test coverage (travis.yml should have the commands).

Yeah, I'm watching the travis builds, and currently confused why it's failing on python 3.5 and python 2. Working on python 3.7
https://travis-ci.org/github/JohannesBuchner/imagehash/jobs/739273952

File: imagehash.py (hunk 2 above)
Comment (joshcoales) [reply imagehash.py:404]: Good point!
There's one in whash too, line 307. Should I fix that?

File: imagehash.py (hunk 1 above)
Comment (joshcoales) [reply imagehash.py:395]: Hmm, good point. In which case, it's just `not self.matches(other)`
I was hoping adding __ne__() would fix the 3.5 and 2.7 errors I see on travis, but it seems not.

File: imagehash.py (hunk 2 above)
Comment (JohannesBuchner) [reply imagehash.py]: Yes, change "255" to "255."

Comment (joshcoales) [reply imagehash.py]: Alright, I've fixed that too. But travis is taking ages (an hour!?) to actually start the tests. And running python 3.5 locally doesn't seem to show the same failure as Travis has in the last few runs

**Comment by joshcoales** [comment]
Installed python 2.7 locally and tested, reproduced the failure, and fixed it. Waiting on a travis run now, and hoping it's green.
https://travis-ci.org/github/JohannesBuchner/imagehash/builds/739404140
Not certain why travis is currently queuing jobs for about an hour before they get run though

**Comment by joshcoales** [comment]
Travis run is green now!

**Comment by joshcoales** [comment]
I'm not liking this actually.. Something is a bit wrong here.
I made a test in this branch that hashes are static between python versions: https://github.com/joshcoales/imagehash/tree/hash-static-test and I can see currently that there's very slightly different segmentation happening between python2.7 and 3.5 at least.. Only a few pixels off, but something's weird there. I'm still investigating.

I'm sorry this has been a bit of a mess. I really hope I can get it all ironed out soon

**Comment by JohannesBuchner** [comment]
Perhaps you can add some unit tests that test that the hash is a certain, hard-coded value (also for the other implementations). This will ensure stability over python version changes and over the evolution of the library.

**Comment by joshcoales** [comment]
Yes, I've tried that here:
https://github.com/joshcoales/imagehash/blob/hash-static-test/tests/test_hash_is_constant.py
And the crop resistant hash fails it, because the segmentation fails to be identical, because the image is slightly different after conversion to greyscale in python3 vs python2:
```
self = <tests.test_hash_is_constant.Test testMethod=test_crop_resistant_segmentation>

    def test_crop_resistant_segmentation(self):
    	# Image pre-processing
    	self.assertEqual(hashlib.md5(self.peppers.tobytes()).hexdigest(), "7c19bbf1c9184471ebb2e7e0086910c6")
    	image = self.peppers.convert("L")
>   	self.assertEqual(hashlib.md5(image.tobytes()).hexdigest(), "61442e74c83cfea67d182481c24c5f3e")

Expected :61db06218cc8b9aba14812d965869120
Actual   :61442e74c83cfea67d182481c24c5f3e
```

**Comment by JohannesBuchner** [comment]
That's odd. I would expect convert("L") is what the hash function does first, so applying this before calling it shouldn't change the output. 

That said, perhaps the version of pillow (and other libraries) is different in python2.7, and behaves differently?

**Comment by joshcoales** [comment]
Yeah, my guess is that the other hashes are all dealing with very small images after converting and resizing, while this one deals with a 300x300px image, so the small errors are not shrunken away?

Could be the Pillow version though. Looks like my py2.7 venv uses pillow 6.2.2, my py3.5 venv uses pillow 7.2.0 and my py3.7 venv is using pillow 7.2.0 also

**Comment by JohannesBuchner** [comment]
"Fix rounding error on RGB to L conversion #4320 [homm]" https://github.com/python-pillow/Pillow/blob/master/CHANGES.rst

**Comment by JohannesBuchner** [comment]
Maybe in that case check that the hash is in the set of the two hash strings.

**Comment by joshcoales** [comment]
Ah ha! I was trying to find the culprit, but that's definitely it. Okay, I'll get those tests handling that difference, and probably add a warning to the crop resistant hash method

**Comment by joshcoales** [comment]
Should I add that hash consistency test to this PR also?
https://github.com/joshcoales/imagehash/blob/hash-static-test/tests/test_hash_is_constant.py

**Comment by joshcoales** [comment]
Hey, what did you think of adding the hash consistency test I linked? Shall we move towards getting this all done?

File: imagehash.py [hunk 3]
```diff
@@ -390,10 +390,21 @@ def __eq__(self, other):
...
+		return not self.matches(other)
+
 	def __sub__(self, other, hamming_cutoff=None, bit_error_rate=None):
 		matches, sum_distance = self.hash_diff(other, hamming_cutoff, bit_error_rate)
+		max_difference = len(self.segment_hashes)
+		if matches == 0:
+			return max_difference
 		max_distance = matches * len(self.segment_hashes[0])
-		return matches + (1 - (sum_distance / max_distance))
+		tie_breaker = 1 - (float(sum_distance) / max_distance)
```
Comment (JohannesBuchner) [inline imagehash.py:402]: Presumably sum_distance can take values between 0 and max_distance, and therefore tie_breaker is between 0 and 1.
When matches == max_difference - 1, and sum_distance == 0, `__sub__` returns zero. Is that good?

File: imagehash.py [hunk 4]
```diff
@@ -495,6 +506,8 @@ def _find_region(remaining_pixels, segmented_pixels):
 def _find_all_segments(pixels, segment_threshold, min_segment_size):
 	"""
 	Finds all the regions within an image pixel array, and returns a list of the regions.
+	WARNING: You will get slightly different segmentation between pillow version 6 and version 7, due to a rounding
```
Comment (JohannesBuchner) [inline imagehash.py:509]: Add a newline before the warning, and change to:
`Note: Slightly different segmentations are produced when using pillow version 6 vs. >=7, due to a change in rounding in the greyscale conversion.`

**Comment by JohannesBuchner** [comment]
> Hey, what did you think of adding the hash consistency test I linked? Shall we move towards getting this all done?

Yes, that looks good!

File: imagehash.py (hunk 3 above)
Comment (joshcoales) [reply imagehash.py:402]: That.. is not good.
If an image matches all segment hashes, so matches = max_difference, and sum_distance == 0, __sub__ returns -1. Very not right.

That `tie_breaker = 1 - ...` should be `tie_breaker = 0 - ...`

Then tie_breaker is between -1 and 0. That seems fine
I'll change that out.

File: imagehash.py (hunk 4 above)
Comment (joshcoales) [reply imagehash.py]: Sounds good, swapped to that wording in 2b9f2e1e6b5d7e891b0c139e8de923452b29908e

**Comment by joshcoales** [comment]
> 
> 
> > Hey, what did you think of adding the hash consistency test I linked? Shall we move towards getting this all done?
> 
> Yes, that looks good!

Awesome, merged that into this branch also then!

**Comment by JohannesBuchner** [comment]
I released a new version of imagehash (4.2) with this merged. Please test it.

I also initiated the travis-ci.com migration.

**Comment by JohannesBuchner** [comment]
@joshcoales Many, many thanks for your work. If you are interested to add more algorithms, they are more than welcome.
Here are some ideas: https://docs.opencv.org/3.4/d4/d93/group__img__hash.html, https://www.phash.org/ and merging back from https://github.com/thorn-oss/perception, https://pippy360.github.io/transformationInvariantImageSearch/.

**Pull request merged**